src/Features/LanguageServer/Protocol/LspMiscellaneousFilesWorkspace.cs

```diff
@@ -0,0 +1,111 @@
+{
+    /// <summary>
+    /// Defines a default workspace for opened LSP files that are not found in any
+    /// workspace registered by the <see cref="ILspWorkspaceRegistrationService"/>.
+    /// If a document added here is subsequently found in a registered workspace, 
+    /// the document is removed from this workspace.
+    /// 
+    /// Future work for this workspace includes supporting basic metadata references (mscorlib, System dlls, etc),
+    /// but that is dependent on having a x-plat mechanism for retrieving those references from the framework / sdk.
+    /// </summary>
+    internal class LspMiscellaneousFilesWorkspace : Workspace
+    {
+        private const string LspMiscellaneousFilesWorkspaceKind = nameof(LspMiscellaneousFilesWorkspaceKind);
+
+        private static readonly LanguageInformation s_csharpLanguageInformation = new(LanguageNames.CSharp, ".csx");
+        private static readonly LanguageInformation s_vbLanguageInformation = new(LanguageNames.VisualBasic, ".vbx");
+
+        private static readonly Dictionary<string, LanguageInformation> s_extensionToLanguageInformation = new()
+        {
+            { ".cs", s_csharpLanguageInformation },
+            { ".csx", s_csharpLanguageInformation },
+            { ".vb", s_vbLanguageInformation },
+            { ".vbx", s_vbLanguageInformation },
+        };
+
+        private readonly ILspLogger _logger;
+
+        public LspMiscellaneousFilesWorkspace(ILspLogger logger) : base(MefHostServices.DefaultHost, LspMiscellaneousFilesWorkspaceKind)
+        {
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Takes in a file URI and text and creates a misc project and document for the file.
+        /// </summary>
+        public Document? AddMiscellaneousDocument(Uri uri)
+        {
+            var uriAbsolutePath = uri.AbsolutePath;
+            if (!s_extensionToLanguageInformation.TryGetValue(Path.GetExtension(uriAbsolutePath), out var languageInformation))
+            {
+                // Only log a warning here since throwing here could take down the LSP server.
+                _logger.TraceError($"Could not find language information for {uri} with absolute path {uriAbsolutePath}");
+                return null;
+            }
+
+            // Create an empty text loader.  The document text is tracked by LSP separately
+            // and forked with the actual LSP text before answering requests, so we can just keep it in the workspace as an empty file.
+            var sourceTextLoader = new SourceTextLoader(SourceText.From(string.Empty), uriAbsolutePath);
+
+            var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument(uri.AbsolutePath, sourceTextLoader, languageInformation, Services, ImmutableArray<MetadataReference>.Empty);
+            OnProjectAdded(projectInfo);
+
+            var id = projectInfo.Documents.Single().Id;
+            return CurrentSolution.GetDocument(id);
```

Comment:
```suggestion
            return CurrentSolution.GetRequiredDocument(id);
```

Your method is marked nullable but we really don't expect that to fail.